### DISCUSSION TITLE
How to do event handling for android and iOS in react native's new architecture for fabric components?

### DISCUSSION THREAD

**Discussion by BraveEvidence** [software-mansion/react-native-screens#1529] [General]
I am trying to learn Fabric for the new architecture . Can someone please help me with this https://stackoverflow.com/questions/72981626/how-to-do-event-handling-for-android-and-ios-in-react-natives-new-architecture. I was trying to go through this repo on how this lib implemented it but found it extremely challenging due to the codebase.

**Comment by kkafar** [comment]
# iOS

Assuming native spec you posted is correct (I'm not really familiar with TypeScript + Codegen duo) when you run `pod install` command in your `ios/` directory [codegen](https://search.brave.com/search?q=react-native+codegen&source=desktop) should generate `EventEmitters` cpp file (+ header) in `React-Codegen` pod (you can see below how to find it in XCode)

[link to the screenshot with Pods structure][1]

You can then use these classes in your `Objective-C++` code to emit the events:

```objective-c++
if (_eventEmitter != nullptr) {
    std::dynamic_pointer_cast<const facebook::react::$$NAME OF YOUR EVENT EMITTER TYPE$$>(_eventEmitter)
        ->onEvent(facebook::react::$$NAME OF YOUR EVENT EMITTER TYPE::OnEvent{.value = $$VALUE$$});
}
```

See links below to see how it is done in [`react-native-screens`](https://github.com/software-mansion/react-native-screens)

1. [Native component spec in `Flow`](https://github.com/software-mansion/react-native-screens/blob/main/src/fabric/ScreenNativeComponent.js)
2. [Event emission in Obj-C++ code](https://github.com/software-mansion/react-native-screens/blob/cc5de7ba28e90ecce9bbb3464ef0a625eaae032f/ios/RNSScreen.mm#L316-L334)
3. [Event exposure in screen manager](https://github.com/software-mansion/react-native-screens/blob/cc5de7ba28e90ecce9bbb3464ef0a625eaae032f/ios/RNSScreen.mm#L1175) - I'm not quite sure if this line is necessary, haven't tested it out yet.

# Android

1. To trigger codegen just run native build. 
2. Define native event class as [done here](https://github.com/software-mansion/react-native-screens/blob/main/android/src/main/java/com/swmansion/rnscreens/events/ScreenAppearEvent.kt)
3. Take a look here for [how to dispatch event](https://github.com/software-mansion/react-native-screens/blob/ce84df60227a02e1e505e1a494ea1ca829788d58/android/src/main/java/com/swmansion/rnscreens/ScreenFragment.kt#L198-L215)
4. [Export your event types](https://github.com/software-mansion/react-native-screens/blob/ce84df60227a02e1e505e1a494ea1ca829788d58/android/src/main/java/com/swmansion/rnscreens/ScreenViewManager.kt#L164-L181)



  [1]: https://i.stack.imgur.com/x52mg.png

**Comment by BraveEvidence** [comment]
@kkafar Thanks will check on this. Do you have any idea on how do it for android. I have tried it for android as I have little experience with it but I am not able to transmit click listener event from native side to js side.  In case you have time check this https://github.com/PritishSawant/ReactNativeFabricEventListenerExample

**Reply by kkafar** [reply]
I updated my previous comment to include links to example implementations for Android. Let me know if you succeeded. In case not, I will follow with more detailed explanation.

**Comment by BraveEvidence** [comment]
<img width="921" alt="Screenshot 2022-07-18 at 7 11 23 PM" src="https://user-images.githubusercontent.com/24708929/179524407-1a62b3f5-358d-4ea2-babd-0ed9548a5422.png">
dispatch is deprecated. What's the alternative?

**Reply by kkafar** [reply]
For this moment, we're sticking to `dispatch` method. If I do some research, I'll let you know.

**Comment by BraveEvidence** [comment]
<img width="971" alt="Screenshot 2022-07-18 at 7 48 36 PM" src="https://user-images.githubusercontent.com/24708929/179531927-77400b35-1a96-4bf2-84af-c46871d8880e.png">
Can you help me in understanding how to refactor the above highlighted code according to my use case?

**Reply by kkafar** [reply]
As I described in main answer: 

```objective-c++
if (_eventEmitter != nullptr) {
    std::dynamic_pointer_cast<const facebook::react::$$NAME OF YOUR EVENT EMITTER TYPE$$>(_eventEmitter)
        ->onEvent(facebook::react::$$NAME OF YOUR EVENT EMITTER TYPE::OnEvent{.value = $$VALUE$$});
}
```

You should locate your event emitters in `EventEmitters.{h,cpp}` files (which you should be able to find following instructions from main answer).

**Comment by BraveEvidence** [comment]
@kkafar Brillaint stuff. Working for ios. Will try for android as well and get back to you

**Comment by BraveEvidence** [comment]
@kkafar I am facing issue on android. I get below screen when I click on my button.
I have pushed the latest code to my repo. Can you please check https://github.com/PritishSawant/ReactNativeFabricEventListenerExample

I did override `dispatch` but both `dispatch` and `dispatchModern` are deprecated. Not sure what the alternative is. 

I also override `getEventData()` but not sure what exactly I need to pass in there as I don't see `getEventData()` overriden
 in the sample you mentioned [here](https://github.com/software-mansion/react-native-screens/blob/main/android/src/main/java/com/swmansion/rnscreens/events/ScreenAppearEvent.kt)

<img width="455" alt="Screenshot 2022-07-19 at 12 50 18 PM" src="https://user-images.githubusercontent.com/24708929/179690551-9a01d862-5c22-4394-8882-0257f72bc54d.png">

**Reply by BraveEvidence** [reply]
Ok I was able to figure it out and it works but I still don't understand few things. 

1. I have to change the event name from `OnClickHandler` to `topOnClickHandler` in Native Android side. 
2. Why do I have to override `getEventData()`
3. What is the alternative to `dispatch` and `RCTEventEmitter` if both are deprecated?

**Reply by BraveEvidence** [reply]
Further more debugging, I think I need to override `getEventData()` as according to my spec file in `TS` I need to return a `string` so may be that is necessary. Not 100% sure.
